xonsh/environ.py

```diff
@@ -7,62 +7,188 @@
 import builtins
 import platform
 import subprocess
+from datetime import datetime
+from functools import partial
 from warnings import warn
 
 from xonsh.tools import TERM_COLORS
 
-def current_branch(cwd=None):
-    """Gets the branch for a current working directory. Returns None
-    if the cwd is not a repository.  This currently only works for git, 
-    bust should be extended in the future.
-    """
-    branch = None
-    cwd = os.getcwd() if cwd is None else cwd
+class PromptFormatter(dict):
+    '''
+    Base class that implements utilities for PromptFormatters.
 
-    # step out completely if git is not installed
-    try:
-        binary_location = subprocess.check_output(['which', 'git'], cwd=cwd,
-                                    stderr=subprocess.PIPE,
-                                    universal_newlines=True)
-        if not binary_location:
-            return branch
-    except subprocess.CalledProcessError:
-        return branch
+    You should read the docs for this class to learn how to write your own
+    Prompt Formatter.  However, you most likely want to inherit from
+    DefaultPromptFormatter if you are writing a custom formatter.
+
+    Custom Prompt Formatters will need to implement their own __init__()
+    method that adds entries to several dicts depending on the needs of the
+    variables they add.  The Formatter may also implement methods to implement
+    those variables (although very simple variables can be written with
+    pre-existing functions.)
+
+    Example:
```

Comment:
sphinx has to run the code through the python compile step so we cannot use a version of sphinx-build running on python2.  Must run it via python3.  I noticed that your scripts/xonsh also used a /usr/bin/python shebang  instead of /usr/bin/python3 so maybe you have your system setup so that python maps to python3 (perhaps just when running certain applications)?  That would explain why you don't need that but other people do (I had to change the shebang line to run from a checkout as well.  The shebang is properly rewritten when installed if python3 seutp.py install is used).

Looks like python3 support in cloud_sptheme on pypi is broken.  They have it fixed in source control but they haven't made a release since 2013 so they haven't pushed the fix out where it's easily available.  I've locally my copy so that it runs.

autodoc isn't doing a good job of organizing the docs for the new classes.  https://toshio.fedorapeople.org/environ.html  what do you think of this plan:

* Pull the formatters out of environ into their own file.
* Use autodoc of specific methods of the formatter classes instead of autodoc of the whole file.  * Turn off inherited-members for the new formatter documentation.
* Either create a new rst page in api for prompt formatters or add it as a separate section o nteh environ page (sine format_prompt() is there).

Question: Do you still want documentation of the prompt variables in format_prompt?  I think it may be more logical to document that in DefaultPromptFormatter (probably in the class docstring since not all format variables are implemented via a method.)  Will need to point the link from the tutorial section to DefaultPromptFormatter as well.